File: Programmers/Level-1/4.py
def solution(numbers, hand):
    answer = ''
    board = {1:(0,0), 2:(0,1), 3:(0,2),
            4:(1,0), 5:(1,1), 6:(1,2),
            7:(2,0), 8:(2,1), 9:(2,2),
            '*':(3,0), 0:(3,1), '#':(3,2)}
    
    left = [1, 4, 7]
    right = [3, 6, 9]
    lhand = '*'
    rhand = '#'
    
    for i in numbers:
        if i in left:
            answer += 'L'
            lhand = i
        elif i in right:
            answer += 'R'
            rhand = i
        else:
            ldis = abs(board[lhand][0]-board[i][0]) + abs(board[lhand][1]-board[i][1])
            rdis = abs(board[rhand][0]-board[i][0]) + abs(board[rhand][1]-board[i][1])
            
            if ldis < rdis:
                answer += 'L'
                lhand = i
            elif rdis < ldis:
                answer += 'R'
                rhand = i
            else:
                if hand == 'right':
                    answer += 'R'
                    rhand = i
                else:
                    answer += 'L'
                    lhand = i
            
    return answer

File: Programmers/Level-1/4.cpp
#include <string>
#include <vector>

using namespace std;

typedef struct {
    int x, y;
} Point;

int distance(Point handloc, Point num) {
    return abs(handloc.x - num.x) + abs(handloc.y - num.y);
}

string solution(vector<int> numbers, string hand) {
    string answer = "";
    Point lhand = { 3,0 };//왼손 시작점
    Point rhand = { 3,2 };//오른손 시작점

    for (int i = 0; i < numbers.size(); i++) {
        int num = numbers[i];
        switch(num) {
        case 1:
            lhand = { 0,0 };
            answer += 'L';
            break;
        case 3:
            rhand = { 0,2 };
            answer += 'R';
            break;
        case 4:
            lhand = { 1,0 };
            answer += 'L';
            break;
        case 6:
            rhand = { 1,2 };
            answer += 'R';
            break;
        case 7:
            lhand = { 2,0 };
            answer += 'L';
            break;
        case 9:
            rhand = { 2,2 };
            answer += 'R';
            break;
        default:
            Point temp;
            if (num == 0) {
                temp = { 3,1 };
            }
            else {
                temp = { num / 3, num % 3 - 1 };
            }

            int ldis = distance(lhand, temp);
            int rdis = distance(rhand, temp);

            if (ldis > rdis) {
                rhand = temp;
                answer += 'R';
            }
            else if (ldis < rdis) {
                lhand = temp;
                answer += 'L';
            }
            else {
                if (hand == "right") {
                    rhand = temp;
                    answer += 'R';
                }
                else {
                    lhand = temp;
                    answer += 'L';
                }                    
            }

            break;
        }
    }
    return answer;
}
